## Exercise 6 - XML

1 Add the necessary libraries (xml and pandas).

In [10]:
import xml.etree.ElementTree as e6
tree = e6.parse("users_first.xml")
root = tree.getroot()

import pandas as pd6
users_collection = []
for i in root:
    users_obj = {"type": i.attrib["type"]}  
    for j in i:
        if j.tag == "address":  
            for k in j:
                users_obj[k.tag] = k.text
        else:
            users_obj[j.tag] = j.text  
    users_collection.append(users_obj)
users_df = pd6.DataFrame(users_collection)
users_df
    

type               name           phone  \
0   customer       Morgan Sykes  1-183-546-6564   
1   employee       Zia Stafford  (621) 483-5714   
2   employee     Cooper Ramirez  (716) 895-3353   
3   customer  Richard Armstrong  (450) 317-2624   
4   employee     Flavia Delaney            None   
5   customer  Branden Henderson  1-314-584-1088   
6   employee    Hector Kirkland  (325) 535-9750   
7   customer        Luke Hooper            None   
8   employee      Moses Dejesus  1-337-767-2455   
9   customer          Finn Pena  1-746-189-1226   
10  employee    Lester Espinoza  1-378-721-6318   
11  employee     Nelle Stafford  1-246-844-4648   
12  customer        Amber Casey            None   
13  customer      Jordan Strong  1-188-391-2011   
14  employee    Elton Mcpherson  1-733-172-5377   

                                         email  \
0             lorem.auctor@vulputatemauris.net   
1                ornare.fusce@felisdonec.co.uk   
2   suspendisse@consectetueradipiscingelit.edu   
3                      lacinia.mattis@urna.org   
4                  rutrum.lorem.ac@inmagna.org   
5                  iaculis@felisnullatempor.ca   
6                           elit@orcilacus.org   
7                vitae.sodales@vitaesodales.ca   
8                      pede@donecdignissim.org   
9                          et.euismod.et@ac.ca   
10                     vel.nisl.quisque@ad.org   
11                          ligula@posuere.com   
12                        diam@nequeetnunc.com   
13                 tincidunt@quamcurabitur.net   
14                     praesent@tempuseu.co.uk   

                        streetaddress             country    postalZip  
0                     2594 Tellus St.  Russian Federation       584585  
1               Ap #419-5905 Nibh Rd.         Netherlands         7308  
2                 667-4582 Mollis Ave               India        43714  
3                Ap #120-495 Ut, Road  Russian Federation        53415  
4                    870-3073 Et, Av.             Austria        03247  
5   P.O. Box 961, 9477 Gravida Avenue            Pakistan        50217  
6              541-4061 Ultricies Ave      United Kingdom         3865  
7           Ap #385-5179 Luctus, Road             Germany         3660  
8          Ap #639-9807 Phasellus St.             Germany       903866  
9               Ap #668-4837 Nec Road         Netherlands        32910  
10                     4395 Elit, St.               Chile  07740-64584  
11          Ap #640-4104 Maecenas St.               Chile       800308  
12             Ap #944-5570 Nunc. Rd.            Pakistan       154758  
13            Ap #426-2862 Lectus Rd.       United States     37144455  
14               110-7294 Tempus Road             Belgium        89525

2 Save the XML structure shown below as a string in a variable. In addition, load the content from the file *users_first.xml* into another variable.

In [8]:
xml_structure = "<?xml version='1.0' encoding='UTF-8'?><records><record type='customer'><name>Lawrence Burke</name><phone>(815) 571-8746</phone><email>vulputate.lacus@elitfermentum.net</email><address><streetaddress>P.O. Box 478, 5254 Mi St.</streetaddress><country>Mexico</country><postalZip>22044</postalZip></address></record><record type='employee'><name>Angela Hinton</name><phone>(997) 412-6965</phone><email>elit.dictum@egestas.net</email><address><streetaddress>696-4409 Nunc Rd.</streetaddress><country>Germany</country><postalZip>661217</postalZip></address></record><record type='customer'><name>Whitney Flowers</name><phone>1-573-471-9738</phone><email>consectetuer.adipiscing.elit@ridiculusmus.ca</email><address><streetaddress>6978 Et, Av.</streetaddress><country>Indonesia</country><postalZip>28164</postalZip></address></record><record type='customer'><name>Darius Burt</name><phone>1-675-212-4934</phone><email>sociis.natoque@ut.org</email><address><streetaddress>7705 Ut Street</streetaddress><country>Indonesia</country><postalZip>94783-300</postalZip></address></record><record type='customer'><name>Solomon Blair</name><phone></phone><email>pede.suspendisse.dui@commodoat.edu</email><address><streetaddress>Ap #877-8946 Sociosqu Street</streetaddress><country>Germany</country><postalZip>113513</postalZip></address></record></records>"

In [7]:
with open("users_first.xml", "r") as file:
    users_first_xml = file.read()

3 From the contents of both variables, form lists C1 and C2, in which the contents of the XML data are stored as dictionary-type objects. **NOTE**: the attributes presented at the record level must be included in the objects and the identifiers under the address identifier must be brought to a higher level next to user's other properties!

Below is a model showing the structure of the first object:

<pre>[{
    'type': 'customer',
    'name': 'Morgan Sykes',
    'phone': '1-183-546-6564',
    'email': 'lorem.auctor@vulputatemauris.net',
    'streetaddress' : '2594 Tellus St.',
    'country': 'Russian Federation',
    'postalZip': '584585'
}]<pre>



In [17]:
C1 = []
C2 = []

for record in root.findall('record'):
    entry = {
        'type': record.attrib['type'],
        'name': record.find('name').text,
        'phone': record.find('phone').text if record.find('phone').text else '',
        'email': record.find('email').text,
        'streetaddress': record.find('address/streetaddress').text,
        'country': record.find('address/country').text,
        'postalZip': record.find('address/postalZip').text,
    }

    if entry['type'] == 'customer':
        C1.append(entry)
    else:
        C2.append(entry)

C1, C2

([{'type': 'customer',
   'name': 'Morgan Sykes',
   'phone': '1-183-546-6564',
   'email': 'lorem.auctor@vulputatemauris.net',
   'streetaddress': '2594 Tellus St.',
   'country': 'Russian Federation',
   'postalZip': '584585'},
  {'type': 'customer',
   'name': 'Richard Armstrong',
   'phone': '(450) 317-2624',
   'email': 'lacinia.mattis@urna.org',
   'streetaddress': 'Ap #120-495 Ut, Road',
   'country': 'Russian Federation',
   'postalZip': '53415'},
  {'type': 'customer',
   'name': 'Branden Henderson',
   'phone': '1-314-584-1088',
   'email': 'iaculis@felisnullatempor.ca',
   'streetaddress': 'P.O. Box 961, 9477 Gravida Avenue',
   'country': 'Pakistan',
   'postalZip': '50217'},
  {'type': 'customer',
   'name': 'Luke Hooper',
   'phone': '',
   'email': 'vitae.sodales@vitaesodales.ca',
   'streetaddress': 'Ap #385-5179 Luctus, Road',
   'country': 'Germany',
   'postalZip': '3660'},
  {'type': 'customer',
   'name': 'Finn Pena',
   'phone': '1-746-189-1226',
   'email': 'et.e

4 Convert both lists into dataframes and finally combine them into one dataframe (remember to reset the index of the new dataframe!). Print the content of the generated dataframe.

In [12]:
import pandas as pd

df_C1 = pd.DataFrame(C1)
df_C2 = pd.DataFrame(C2)

df_combined = pd.concat([df_C1, df_C2], ignore_index=True)

df_combined


type               name           phone  \
0   customer       Morgan Sykes  1-183-546-6564   
1   customer  Richard Armstrong  (450) 317-2624   
2   customer  Branden Henderson  1-314-584-1088   
3   customer        Luke Hooper                   
4   customer          Finn Pena  1-746-189-1226   
5   customer        Amber Casey                   
6   customer      Jordan Strong  1-188-391-2011   
7   employee       Zia Stafford  (621) 483-5714   
8   employee     Cooper Ramirez  (716) 895-3353   
9   employee     Flavia Delaney                   
10  employee    Hector Kirkland  (325) 535-9750   
11  employee      Moses Dejesus  1-337-767-2455   
12  employee    Lester Espinoza  1-378-721-6318   
13  employee     Nelle Stafford  1-246-844-4648   
14  employee    Elton Mcpherson  1-733-172-5377   

                                         email  \
0             lorem.auctor@vulputatemauris.net   
1                      lacinia.mattis@urna.org   
2                  iaculis@felisnullatempor.ca   
3                vitae.sodales@vitaesodales.ca   
4                          et.euismod.et@ac.ca   
5                         diam@nequeetnunc.com   
6                  tincidunt@quamcurabitur.net   
7                ornare.fusce@felisdonec.co.uk   
8   suspendisse@consectetueradipiscingelit.edu   
9                  rutrum.lorem.ac@inmagna.org   
10                          elit@orcilacus.org   
11                     pede@donecdignissim.org   
12                     vel.nisl.quisque@ad.org   
13                          ligula@posuere.com   
14                     praesent@tempuseu.co.uk   

                        streetaddress             country    postalZip  
0                     2594 Tellus St.  Russian Federation       584585  
1                Ap #120-495 Ut, Road  Russian Federation        53415  
2   P.O. Box 961, 9477 Gravida Avenue            Pakistan        50217  
3           Ap #385-5179 Luctus, Road             Germany         3660  
4               Ap #668-4837 Nec Road         Netherlands        32910  
5              Ap #944-5570 Nunc. Rd.            Pakistan       154758  
6             Ap #426-2862 Lectus Rd.       United States     37144455  
7               Ap #419-5905 Nibh Rd.         Netherlands         7308  
8                 667-4582 Mollis Ave               India        43714  
9                    870-3073 Et, Av.             Austria        03247  
10             541-4061 Ultricies Ave      United Kingdom         3865  
11         Ap #639-9807 Phasellus St.             Germany       903866  
12                     4395 Elit, St.               Chile  07740-64584  
13          Ap #640-4104 Maecenas St.               Chile       800308  
14               110-7294 Tempus Road             Belgium        89525

5 Some users are missing a phone number. Print the number of these users.

In [18]:
missing_phone_count = df_combined[df_combined['phone'] == ''].shape[0]
missing_phone_count


3